### PR TITLE
[actsync] sync blob action

### PR DESCRIPTION
**Pull request by envestcc**
# Description

This modification is made to support the efficient synchronization of the blob action introduced in EIP-4844 on the network. Specific changes include: 
- Introducing an action synchronization mechanism for big action that broadcasts transaction hashes instead of transactions themselves to reduce network overhead
- Configuring the blob action to use this mechanism for synchronization

Based on #4232 

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: actsync/actionsync.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	if err := as.TurnOff(); err != nil {
+		return err
+	}
+	close(as.syncChan)
+	as.wg.Wait()
+	return nil
+}
+
+// RequestAction requests an action by hash
+func (as *ActionSync) RequestAction(ctx context.Context, hash hash.Hash256) error {
```
Comment (CoderZhi) [inline actsync/actionsync.go:77]: no error will be returned in this function

File: actsync/actionsync.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		repeat = len(neighbors)
+	}
+	peers := make([]peer.AddrInfo, repeat)
+	for i := 0; i < repeat; i++ {
+		peer := neighbors[fastrand.Uint32n(uint32(len(neighbors)))]
+		peers[i] = peer
+	}
+	if len(peers) == 0 {
+		return nil, errors.New("no peers")
+	}
```
Comment (CoderZhi) [inline actsync/actionsync.go:163]:
```
if repeat == 0 {
  ...?
}
```

**Review by CoderZhi — APPROVED** [review]

File: api/coreservice.go [hunk 3]
```diff
@@ -497,14 +497,23 @@ func (core *coreService) SendAction(ctx context.Context, in *iotextypes.Action)
 		return "", st.Err()
 	}
 	// If there is no error putting into local actpool, broadcast it to the network
-	if core.messageBatcher != nil {
+	// TODO: broadcast action hash if it's blobTx
+	isBlobTx := false
+	out := proto.Message(in)
```
Comment (dustinxie) [inline api/coreservice.go:502]: why need to add `proto.Message()`

File: dispatcher/msg_queue.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
+package dispatcher
```
Comment (dustinxie) [inline dispatcher/msg_queue.go:1]: merge PR4232 so this file is added

File: dispatcher/dispatcher_test.go [hunk 5]
```diff
@@ -108,3 +199,64 @@ func (ds *dummySubscriber) HandleNodeInfoRequest(context.Context, peer.AddrInfo,
...
+type counterSubscriber struct {
+	block       atomic.Int32
+	blockSync   atomic.Int32
+	action      atomic.Int32
+	consensus   atomic.Int32
+	nodeInfo    atomic.Int32
+	nodeInfoReq atomic.Int32
+	actionReq   atomic.Int32
+	actionHash  atomic.Int32
+}
```
Comment (dustinxie) [inline dispatcher/dispatcher_test.go:220]: this is for testing purpose?

File: go.mod [hunk 6]
```diff
@@ -251,3 +251,5 @@ require (
 replace github.com/ethereum/go-ethereum => github.com/iotexproject/go-ethereum v1.7.4-0.20240321181400-fc0405c7567c
 
 replace golang.org/x/xerrors => golang.org/x/xerrors v0.0.0-20190212162355-a5947ffaace3
+
+replace github.com/iotexproject/iotex-proto => github.com/envestcc/iotex-proto v0.0.0-20240730104016-c6d74d42763b
```
Comment (dustinxie) [inline go.mod:255]: file a PR to iotex-proto

File: actsync/actionsync.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+func (as *ActionSync) triggerSync() {
+	defer as.wg.Done()
+	ticker := time.NewTicker(as.cfg.Interval)
+	defer ticker.Stop()
+	for {
+		if !as.IsReady() {
+			return
```
Comment (dustinxie) [inline actsync/actionsync.go:127]: is it possible it runs to here before `as.TurnOn()` is run, then will return?

File: dispatcher/subscriber.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// Subscriber is the dispatcher subscriber interface
+type Subscriber interface {
+	ReportFullness(context.Context, iotexrpc.MessageType, float32)
+	HandleAction(context.Context, *iotextypes.Action) error
+	HandleBlock(context.Context, string, *iotextypes.Block) error
+	HandleSyncRequest(context.Context, peer.AddrInfo, *iotexrpc.BlockSync) error
+	HandleConsensusMsg(*iotextypes.ConsensusMessage) error
+	HandleNodeInfoRequest(context.Context, peer.AddrInfo, *iotextypes.NodeInfoRequest) error
+	HandleNodeInfo(context.Context, string, *iotextypes.NodeInfo) error
+	HandleActionRequest(ctx context.Context, peer peer.AddrInfo, actHash hash.Hash256) error
```
Comment (dustinxie) [inline dispatcher/subscriber.go:21]: better name to indicate for special type of action (blob so far)

File: actsync/actionsync.go (hunk 7 above)
Comment (dustinxie) [reply actsync/actionsync.go:127]: for {
   if !as.IsReady() {
       yield(100ms)
   }
}
for {
    // event loop
}

File: chainservice/chainservice.go [hunk 9]
```diff
@@ -105,9 +108,37 @@ func (cs *ChainService) HandleAction(ctx context.Context, actPb *iotextypes.Acti
 	if err != nil {
 		log.L().Debug(err.Error())
 	}
+	hash, err := act.Hash()
+	if err != nil {
+		return err
+	}
+	cs.actionsync.ReceiveAction(ctx, hash)
```
Comment (dustinxie) [inline chainservice/chainservice.go:115]: only do this for unicast?

File: api/coreservice.go (hunk 3 above)
Comment (envestcc) [reply api/coreservice.go:502]: for unifying sending action and action hash

File: dispatcher/dispatcher_test.go (hunk 5 above)
Comment (envestcc) [reply dispatcher/dispatcher_test.go:220]: yes, just for testing

File: actsync/actionsync.go (hunk 7 above)
Comment (envestcc) [reply actsync/actionsync.go:127]: as discussed offline. added a quit chan to handle this

File: dispatcher/subscriber.go (hunk 8 above)
Comment (envestcc) [reply dispatcher/subscriber.go:21]: it is generic mechanism for action sync, not limited only for blob, but we just use blob on it

File: chainservice/chainservice.go (hunk 9 above)
Comment (envestcc) [reply chainservice/chainservice.go:116]: add a TODO, will fix after tx type introduced

File: actsync/actionsync.go [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	l := log.L().With(log.Hex("hash", hash[:]))
+	neighbors, err := as.selectPeers()
+	if err != nil {
+		l.Debug("Failed to get neighbors", zap.Error(err))
+		return err
+	}
+	success := false
+	for i := range neighbors {
+		if err := as.helper.UnicastOutbound(ctx, neighbors[i], &iotexrpc.ActionSync{Hashes: [][]byte{hash[:]}}); err != nil {
+			l.Debug("Failed to request action", zap.Error(err), zap.String("peer", neighbors[i].String()))
```
Comment (dustinxie) [inline actsync/actionsync.go:189]: `ActionSync` is designed to contain multiple hash/actions, so can use 1 message to request more than 1 actions per outbound unicast. Should consider this to enhance the network efficiency?

Comment (envestcc) [reply actsync/actionsync.go:189]: good idea, can impl it in next pr

Comment (dustinxie) [reply actsync/actionsync.go:189]: add a note/TODO

**Review by dustinxie — APPROVED** [review]

**Pull request merged**